### PR TITLE
Improve ICD Storage

### PR DESCRIPTION
**Pull request by yunhanw-google**
--Add ICD GetEntry API to retrieve the ICDClientInfo when user need to get hint and instructions for UserActiveModeTriggerHint
--Add RemoveKey API to delete key when registration fails

### CODE REVIEW

**Review by joonhaengHeo — APPROVED** [review]

File: src/app/icd/client/ICDClientStorage.h [hunk 1]
```diff
@@ -37,22 +36,13 @@ namespace app {
...
-    using ICDClientInfoIterator = CommonIterator<ICDClientInfo>;
-
     virtual ~ICDClientStorage() = default;
 
-    /**
-     * Iterate through persisted ICD Client Info
-     *
-     * @return A valid iterator on success. Use CommonIterator accessor to retrieve ICDClientInfo
-     */
-    virtual ICDClientInfoIterator * IterateICDClientInfo() = 0;
```
Comment (Damian-Nordic) [inline src/app/icd/client/ICDClientStorage.h:49]: So how will a user know after reboot which entries are available if this API is removed?

Comment (yunhanw-google) [reply src/app/icd/client/ICDClientStorage.h]: I may not realize the user case, why application need to iterate entries? in which scenario?

Comment (Damian-Nordic) [reply src/app/icd/client/ICDClientStorage.h]: I may not understand how the API is supposed to be used, but I imagined that after a reboot an ICD would need to recreate ICD sender structures to know where to send the check-in messages. If the `GetEntry` takes a peer node ID, how do I know all the peer node IDs that I should query?

**Review by jmartinez-silabs — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

File: src/app/icd/client/ICDClientStorage.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/icd/client/ICDClientStorage.h]: @Damian-Nordic This is client-side API, not server-side.  The recipient of check-in messages, not the sender.

File: src/app/icd/client/ICDClientStorage.h [hunk 2]
```diff
@@ -61,16 +61,31 @@ class ICDClientStorage
...
      *
-     * @param[in] aICDClientInfo the updated ICD Client Info.
+     * @param[in] clientInfo the updated ICD Client Info.
      */
     virtual CHIP_ERROR StoreEntry(const ICDClientInfo & clientInfo) = 0;
 
+    /**
+     * Remove ICD key from clientInfo when ICD registration fails
+     *
+     * @param[inout] clientInfo the updated ICD Client Info.
```
Comment (bzbarsky-apple) [inline src/app/icd/client/ICDClientStorage.h:71]: Updated in what way?  The one implementation here is not updating anything about the clientinfo, as far as I can tell.

File: src/app/icd/client/ICDClientStorage.h [hunk 3]
```diff
@@ -61,16 +61,31 @@ class ICDClientStorage
...
+     */
+    virtual void RemoveKey(ICDClientInfo & clientInfo) = 0;
+
+    /**
+     * Get ICD ClientInfo from storage
+     * One user case is to retrieve UserActiveModeTriggerHint and inform how user how to wake up sleepy device.
```
Comment (bzbarsky-apple) [inline src/app/icd/client/ICDClientStorage.h:77]: s/user/use/

And

s/inform how user how to wake up sleepy/inform the user how to wake up a sleepy/ (note: multiple grammar fixes here).

File: src/app/icd/client/ICDClientStorage.h [hunk 4]
```diff
@@ -61,16 +61,31 @@ class ICDClientStorage
...
+     * @param[in] peerNode scoped node with peer node id and fabric index
+     * @param[out] clientInfo the ICD Client Info.
+     */
+    virtual CHIP_ERROR GetEntry(const ScopedNodeId & peerNode, ICDClientInfo & clientInfo) = 0;
```
Comment (bzbarsky-apple) [inline src/app/icd/client/ICDClientStorage.h:81]: If there is no info for this node id, what's the expected way for the callee to communicate that?  Needs to be documented.

File: src/app/icd/client/ICDClientStorage.h [hunk 5]
```diff
@@ -61,16 +61,31 @@ class ICDClientStorage
...
+     * @param[in] clientInfo the updated ICD Client Info.
      */
     virtual CHIP_ERROR StoreEntry(const ICDClientInfo & clientInfo) = 0;
 
+    /**
+     * Remove ICD key from clientInfo when ICD registration fails
+     *
+     * @param[inout] clientInfo the updated ICD Client Info.
+     */
+    virtual void RemoveKey(ICDClientInfo & clientInfo) = 0;
```
Comment (bzbarsky-apple) [inline src/app/icd/client/ICDClientStorage.h:73]: What should this do if there is no key for that clientinfo to start with?

File: src/app/icd/client/ICDClientStorage.h (hunk 1 above)
Comment (Damian-Nordic) [reply src/app/icd/client/ICDClientStorage.h]: Ah, sorry for the confusion!

**Pull request merged**